File: reocurring-characters/reoccur.py
def reoccur(s):
    occur = set()
    for i in range(len(s)):
        if s[i] in occur:
            return s[i]
        else:
            occur.add(s[i])
    return False

def reoccur_no_mem(s):
    if len(s) == 0:
        return False
    
    last = 0
    for i in range(len(s)-1):
        for j in range(i+1, len(s)):
            if s[i] == s[j]:
                if last == 0 or j < last:
                    last = j
    
    if last == 0:
        return False
    else:
        return s[last]

File: reocurring-characters/reoccur.go
package main

import "fmt"

// Reoccur detects the first re-occurring rune in a string
func Reoccur(s string) (rune, bool) {
	set := make(map[rune]struct{})
	for _, r := range s {
		if _, ok := set[r]; ok {
			return r, true
		}
		set[r] = struct{}{}
	}

	return 'a', false
}

// ReoccurNoMem detects the first reoccurring character in a string using no memory
func ReoccurNoMem(s string) (rune, bool) {
	if len(s) == 0 {
		return 'z', false
	}

	runes := []rune(s)
	var last int
	for i := 0; i < len(runes)-1; i++ {
		for j := i + 1; j < len(runes); j++ {
			if runes[i] == runes[j] {
				if last == 0 || j < last {
					last = j
				}
			}
		}
	}

	return runes[last], last != 0
}

func main() {
	fmt.Println("Reoccurring runes in a string")
}
